hydra/core/utils.py

```diff
@@ -131,6 +134,7 @@ def run_job(
         with open_dict(config.hydra.runtime):
             config.hydra.runtime.output_dir = os.path.abspath(output_dir)
 
+    # update Hydra config
     HydraConfig.instance().set_config(config)
```

Comment:
I think this one can be deleted.

Reply:
This is still needed since we need  to update the  value of `hydra.runtime.output_dir` - if not updated an example Hydra app like the following won't work as  expected
```python3
@hydra.main(config_path=".", config_name="config")
def my_app(_: DictConfig) -> None:
    print(HydraConfig.instance().get().runtime.output_dir)

```

Reply:
I see! Currently, the core tests are still passing even if the second call to `set_config` is deleted. Could you please add the statement `print(HydraConfig.instance().get().runtime.output_dir)` to the new `my_app.py` test file and check the output?

Reply:
makes sense, added! :) Thanks

Reply:
It feels to me like a hack to call `HydraConfig.instance().set_config` twice.
The root issue in my mind is that `HydraConfig.instance().set_config` is creating a copy that is different from the original object (which is `config.hydra`). I think there should just be one copy of the hydra config.

I have a way to make these two copies share memory, so that either one can be updated without needing to call `HydraConfig.set_config` a second time.
Here is the diff comparing with the latest commit (https://github.com/facebookresearch/hydra/pull/1979/commits/04d7876eb3241cfa352c25e59af96b388d3f67ff) on this PR branch:
```diff
diff --git a/hydra/core/hydra_config.py b/hydra/core/hydra_config.py
index 6ebcca763f..6ba721cf5b 100644
--- a/hydra/core/hydra_config.py
+++ b/hydra/core/hydra_config.py
@@ -1,7 +1,7 @@
 # Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
 from typing import Any, Optional

-from omegaconf import DictConfig, OmegaConf
+from omegaconf import DictConfig, OmegaConf, read_write

 from hydra.conf import HydraConf
 from hydra.core.singleton import Singleton
@@ -23,6 +23,10 @@ class HydraConfig(metaclass=Singleton):
         # into the user's config.
         self.cfg = OmegaConf.masked_copy(cfg, "hydra")  # type: ignore
         self.cfg.hydra._set_parent(cfg)  # type: ignore
+        # Ensure cfg.hydra and self.cfg.hydra refer to the same object:
+        with read_write(cfg):
+            with open_dict(cfg):
+                cfg.hydra = self.get()

     @staticmethod
     def get() -> HydraConf:
diff --git a/hydra/core/utils.py b/hydra/core/utils.py
index 602f7f3a91..e993294521 100644
--- a/hydra/core/utils.py
+++ b/hydra/core/utils.py
@@ -134,8 +134,6 @@ def run_job(
         with open_dict(config.hydra.runtime):
             config.hydra.runtime.output_dir = os.path.abspath(output_dir)

-    # update Hydra config
-    HydraConfig.instance().set_config(config)
     _chdir = None
     try:
         ret = JobReturn()
```
@jieru-hu, what do you think about this idea?
I ran the tests locally and they seem to be passing.

Reply:
Resolved via PMs.

Reply:
resolving per offline discussions cc @Jasha10